### PR TITLE
overload: scale transport socket connect timeout

### PR DESCRIPTION
**Pull request by akonradi**
Commit Message: Scale transport socket connect timeout in response to load
Additional Description:
Add support for scaling the transport socket connect timeout with load.

Risk Level: low
Testing: added tests and ran affected tests
Docs Changes: none
Release Notes: none
Platform Specific Features: none
Fixes: #11426

/assign @KBaichoo

### CODE REVIEW

File: source/server/overload_manager_impl.cc
Comment (KBaichoo) [inline source/server/overload_manager_impl.cc:146]: think you meant `TransportSocketConnectTimeout`?

**Review by KBaichoo — DISMISSED** [review]
LGTM; most of it was refactoring / interface changes.

**Comment by yanavlasov** [comment]
format check failed

File: source/server/overload_manager_impl.cc
Comment (akonradi) [reply source/server/overload_manager_impl.cc]: That's embarrassing, and that there wasn't a test that caught it is problematic. I've added a test to verify that the correct scale factors are applied for the different timer types.

**Review by akonradi — COMMENTED** [review]
Fixed formatting too.

File: include/envoy/thread_local/thread_local.h
Comment (antoniovicente) [inline include/envoy/thread_local/thread_local.h:10]: I think that the dispatcher could be forward declared and dispatcher interface library depedency removed as a dependency if that would make some of these changes easier.

File: include/envoy/event/dispatcher.h
Comment (antoniovicente) [inline include/envoy/event/dispatcher.h:22]: Should you consider forward declaration for Server::ThreadLocalOverloadState ?

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [inline source/common/event/dispatcher_impl.cc:12]: Is this include needed?

File: source/common/event/dispatcher_impl.h
Comment (antoniovicente) [inline source/common/event/dispatcher_impl.h:15]: Is this include needed?

**Review by antoniovicente — COMMENTED** [review]
Great to see this happening.  First set of comments related just to PR size and unrelated looking changes.

File: include/envoy/server/overload/overload_manager.h
Comment (akonradi) [reply include/envoy/server/overload/overload_manager.h:7]: I'm happy to do the refactor in a separate PR. I don't have strong opinions about forward declarations vs `#include`s other than an inclination to follow the style guide.

**Review by akonradi — COMMENTED** [review]
Happy to do the refactoring in a separate PR, or to talk about making the thread-local overload state accessible via the dispatcher.

File: include/envoy/event/dispatcher.h
Comment (antoniovicente) [reply include/envoy/event/dispatcher.h:22]: I am aware of the guidelines at: https://google.github.io/styleguide/cppguide.html#Forward_Declarations .

See "The decision is: Try to avoid forward declarations of entities defined in another project."

The classes involved here are both part of the Envoy project so some of the justifications listed don't fully apply in this case.  I would personally use forward declaration to resolve these circular dependencies.  If you prefer to refactor headers, please do it in a separate PR in order to simplify review of functional changes in this PR.

**Comment by antoniovicente** [comment]
> Happy to do the refactoring in a separate PR, or to talk about making the thread-local overload state accessible via the dispatcher.

See https://github.com/envoyproxy/envoy/pull/13800#discussion_r520149464

If you decide to expose this via dispatcher instead, I suggest you add a createRangeTimer method to the dispatcher interface instead of providing a way to access thread locals via the dispatcher interface.

**Comment by akonradi** [comment]
As discussed offline, I think I'd prefer keeping the scaled timer accessible via the OverloadManager, not the Dispatcher, since the scaling factor is determined by the OM. To that end, I've sent #13999 to do the refactoring requested above separate from this PR.

File: api/envoy/config/overload/v3/overload.proto
Comment (antoniovicente) [inline api/envoy/config/overload/v3/overload.proto:97]: nit: This line seem very long.  Can you rerun format?

File: include/envoy/event/scaled_range_timer_manager.h
Comment (antoniovicente) [reply include/envoy/event/scaled_range_timer_manager.h:60]: It seems very unusual to implement these functions in a different library.  Consider the case of "ForTest" methods.  We tend to implement those in the same library as the main object despite they only being used in tests.  I think that modern linkers are capable of removing unused code from the resulting binaries, but don't know if they do so by default.

File: test/common/network/connection_impl_test.cc
Comment (antoniovicente) [inline test/common/network/connection_impl_test.cc:250]: nit: unusual looking formatting, can you rerun format?

File: test/server/overload_manager_impl_test.cc
Comment (antoniovicente) [inline test/server/overload_manager_impl_test.cc:543]: nit: Please move this list of cases inside the relevant test case to avoid having a large separation between these definitions and the test body.

**Review by antoniovicente — APPROVED** [review]

File: api/envoy/config/overload/v3/overload.proto
Comment (akonradi) [reply api/envoy/config/overload/v3/overload.proto]: Fixed manually since the formatter doesn't seem to touch proto comments.

File: include/envoy/event/scaled_range_timer_manager.h
Comment (akonradi) [reply include/envoy/event/scaled_range_timer_manager.h]: I'm sold, the confusion isn't worth the tiny bit of size gains. Moved these definitions inline.

File: source/common/event/dispatcher_impl.cc
Comment (akonradi) [reply source/common/event/dispatcher_impl.cc]: Neither is needed since thread_local_overload_state.h is included in the interface header.

File: source/common/network/connection_impl.cc
Comment (akonradi) [reply source/common/network/connection_impl.cc:749]: As mentioned elsewhere, I'd rather keep scaled timer creation going through the Overload Manager since it is ultimately responsible for determining the scale factor of those timers. We can revisit making the Dispatcher aware of the OM later.

File: test/common/network/connection_impl_test.cc
Comment (akonradi) [reply test/common/network/connection_impl_test.cc:250]: Reran format.

File: test/server/overload_manager_impl_test.cc
Comment (akonradi) [reply test/server/overload_manager_impl_test.cc:502]: This list matches exactly what kReducedTimeoutsConfig produces, so I wanted to keep them close together. Please push back if that's not a good enough reason.

File: include/envoy/event/scaled_range_timer_manager.h
Comment (antoniovicente) [inline include/envoy/event/scaled_range_timer_manager.h:70]: Would this work as an alternate way to delegate to the (private) base class method?
using absl::variant<ScaledMinimum, AbsoluteMinimum>::operator==;

File: test/server/overload_manager_impl_test.cc
Comment (antoniovicente) [reply test/server/overload_manager_impl_test.cc:502]: I don't know which of the two separations is more problematic.  I would consider having both in the test case so everything is in 1 place, or move OverloadManagerImplTest.TimerTypesProduceCorrectMinimums so it is just below this list definition.

File: test/server/overload_manager_impl_test.cc
Comment (antoniovicente) [inline test/server/overload_manager_impl_test.cc:561]: The previous 2 test cases seem redundant with OverloadManagerImplTest.TimerTypesProduceCorrectMinimums below.

File: include/envoy/event/scaled_range_timer_manager.h
Comment (akonradi) [reply include/envoy/event/scaled_range_timer_manager.h:72]: I think that's the only reason, yeah.

File: include/envoy/event/scaled_range_timer_manager.h
Comment (akonradi) [reply include/envoy/event/scaled_range_timer_manager.h:70]: No, because it's not defined as a method. I hoped that removing the operator== declaration entirely would work but the compiler doesn't find the free operator== function for absl::variant because of the private inheritance.

File: test/server/overload_manager_impl_test.cc
Comment (akonradi) [reply test/server/overload_manager_impl_test.cc:502]: Moved the test case; that's a good option.

File: test/server/overload_manager_impl_test.cc
Comment (akonradi) [reply test/server/overload_manager_impl_test.cc]: Agreed. Removed.

File: include/envoy/event/scaled_range_timer_manager.h
Comment (antoniovicente) [reply include/envoy/event/scaled_range_timer_manager.h:72]: Moving as a data member feels cleaner.  Consider this version:

class ScaledTimerMinimum  {
public:
  // Forward arguments to impl_'s constructor.                                                                                                                                                                                                                                                                                 
  template<typename T> ScaledTimerMinimum(T arg) : impl_(arg) {}

  // Computes the minimum value for a given maximum timeout. If this object was constructed with a                                                                                                                                                                                                                             
  // - ScaledMinimum value:                                                                                                                                                                                                                                                                                                    
  //     the return value is the scale factor applied to the provided maximum.                                                                                                                                                                                                                                                 
  // - AbsoluteMinimum:                                                                                                                                                                                                                                                                                                        
  //     the return value is that minimum, and the provided maximum is ignored.                                                                                                                                                                                                                                                
  std::chrono::milliseconds computeMinimum(std::chrono::milliseconds maximum) const {
    struct Visitor {
      explicit Visitor(std::chrono::milliseconds value) : value_(value) {}
      std::chrono::milliseconds operator()(ScaledMinimum scale_factor) {
        return std::chrono::duration_cast<std::chrono::milliseconds>(scale_factor.scale_factor_ *
                                                                     value_);
      }
      std::chrono::milliseconds operator()(AbsoluteMinimum absolute_value) {
        return absolute_value.value_;
      }
      const std::chrono::milliseconds value_;
    };
    return absl::visit(Visitor(maximum), impl_);
  }

  inline bool operator==(const ScaledTimerMinimum& other) const {
    return impl_ == other.impl_;
  }

  inline friend std::ostream& operator<<(std::ostream& output, const ScaledTimerMinimum& minimum) {
    return absl::visit([&](const auto& minimum) -> std::ostream& { return output << minimum; },
                       minimum.impl_);
  }
 private:
  absl::variant<ScaledMinimum, AbsoluteMinimum> impl_;
};

Comment (akonradi) [reply include/envoy/event/scaled_range_timer_manager.h]: Thanks, done.

**Review by antoniovicente — DISMISSED** [review]

**Comment by akonradi** [comment]
@envoyproxy/api-shepherds can I get a review here? The API change is trivial (add an enum value)

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by antoniovicente** [comment]
Hey @lizan, can you do a pass as owner of core networking and senior maintainer?

**Comment by antoniovicente** [comment]
This PR fixes the rest of Issue #11426

**Comment by akonradi** [comment]
@lizan please review when you get a chance

**Comment by antoniovicente** [comment]
Build failure in CI after merge.

2020-11-24T00:55:37.7906957Z test/common/network/connection_impl_test.cc:361:3: error: no member named 'gmock_createScaledTimer_' in 'Envoy::Server::MockThreadLocalOverloadState'; did you mean 'gmock_createScaledTypedTimer_'?
2020-11-24T00:55:37.7907702Z   EXPECT_CALL(mocks.overload_state_,

/wait

File: test/common/network/connection_impl_test.cc
Comment (nezdolik) [inline test/common/network/connection_impl_test.cc:360]: i wonder if this would result into memory leak, wrap in std::make_unique?

Comment (akonradi) [reply test/common/network/connection_impl_test.cc:360]: It gets returned by the EXPECT_CALL on line 363, which wraps it in a unique_ptr.

File: source/common/network/connection_impl.cc
Comment (antoniovicente) [inline source/common/network/connection_impl.cc:749]: I did a local change that undid changes to this line and fixed all the unused variable compile errors, and noticed that the only test that failed //test/common/network:connection_impl_test due to changes to mock expectations.  It would be good to have some e2e integration test coverage for this functionality.

This issue also came up in https://github.com/envoyproxy/envoy/pull/14155 .  We have multiple scaled timers, we should have some generic recipes for how to integration test scaled timer functionality.  Possible strawman drive the connection to a certain state, force proxy into overload and verify that the timeout triggers shortly afterwards.

File: source/common/network/connection_impl.cc
Comment (ggreenway) [reply source/common/network/connection_impl.cc:750]: Instead of plumbing the OM widely, could the dispatcher hold a reference to it? I dislike that there are now two very different places to create timers: OM and dispatcher.

File: include/envoy/event/scaled_timer_minimum.h
Comment (ggreenway) [inline include/envoy/event/scaled_timer_minimum.h:4]: I'm worried that adding this and the related inline methods here may have a detrimental effect on compile-time. Does this need to live in a header?

File: source/server/server.cc
Comment (ggreenway) [inline source/server/server.cc:473]: I don't know if this causes an actual bug or not, but when `this` gets destructed, the `overload_manager_` will be destructed before `handler_`, so `handler_` will have a dangling reference to a destructed object. Re-order class members?

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: source/common/network/connection_impl.cc
Comment (akonradi) [reply source/common/network/connection_impl.cc:749]: Sent https://github.com/envoyproxy/envoy/pull/14290 to do just that for the existing scaled timer.

File: include/envoy/event/scaled_timer_minimum.h
Comment (akonradi) [reply include/envoy/event/scaled_timer_minimum.h:4]: The method definitions could be moved to a .cc file. Should I put that under `include` or move the whole thing to `source/common/common`?

File: source/common/network/connection_impl.cc
Comment (antoniovicente) [reply source/common/network/connection_impl.cc:750]: The dispatcher is also owned by the thread.  I wonder if the solution would be to have the dispatcher own the thread overload state instead of having it be part of the thread local storage.

**Comment by antoniovicente** [comment]
Can you do a master merge and resolve "conflicts"?

**Comment by akonradi** [comment]
Conflicts are resolved, not sure where the format failure is coming from.

**Comment by antoniovicente** [comment]
> Conflicts are resolved, not sure where the format failure is coming from.

Seems like there was an error fetching some python packages.

/retest

2020-12-08T19:28:45.4825077Z ERROR: An error occurred during the fetch of repository 'configs_pip3':
2020-12-08T19:28:45.4826142Z    Traceback (most recent call last):
2020-12-08T19:28:45.4826924Z 	File "/build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/rules_python/python/pip.bzl", line 49, column 13, in _pip_import_impl
2020-12-08T19:28:45.4827927Z 		fail("pip_import failed: %s (%s)" % (result.stdout, result.stderr))
2020-12-08T19:28:45.4829188Z Error in fail: pip_import failed: Collecting Jinja2==2.11.2 (from -r /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/envoy/configs/requirements.txt (line 1))

File: include/envoy/event/scaled_timer_minimum.h
Comment (lizan) [reply include/envoy/event/scaled_timer_minimum.h:4]: These ostream helpers don't have to be defined inside the struct and you don't even need `friend` since the members are public, so we can define in another .h files, where are they used?

File: test/common/network/connection_impl_test.cc
Comment (lizan) [reply test/common/network/connection_impl_test.cc:250]: nit: will explicitly constructing `Server::MockThreadLocalOverloadState()`  make formatter happy?

File: source/common/network/connection_impl.cc
Comment (lizan) [reply source/common/network/connection_impl.cc:772]: > The dispatcher is also owned by the thread. I wonder if the solution would be to have the dispatcher own the thread overload state instead of having it be part of the thread local storage.

Yeah I think that makes sense.

Comment (ggreenway) [reply source/common/network/connection_impl.cc:772]: +1

**Comment by akonradi** [comment]
I'm trying to figure out what integrating the thread-local overload state into the dispatcher would look like. Should I consider that a blocker for getting this PR merged?

**Comment by antoniovicente** [comment]
> I'm trying to figure out what integrating the thread-local overload state into the dispatcher would look like. Should I consider that a blocker for getting this PR merged?

I defer to @ggreenway 's opinion.

Seems like you accumulated some merge conflicts as your other PRs which modify some of the same files got merged in.

File: source/common/network/connection_impl.cc
Comment (akonradi) [reply source/common/network/connection_impl.cc:772]: I started working on this. It's doable but introduces some weirdness; see #14401 for more.

**Comment by ggreenway** [comment]
> I'm trying to figure out what integrating the thread-local overload state into the dispatcher would look like. Should I consider that a blocker for getting this PR merged?

Yes, I'd like to have that sorted out before we merge this.

File: source/common/network/connection_impl.cc
Comment (akonradi) [reply source/common/network/connection_impl.cc:772]: Made a fresh attempt at this in #14679 which feels reasonably clean.

**Comment by akonradi** [comment]
I've merged main. This is now using Dispatcher::createScaledTimer, which obviates the need for plumbing around the ThreadLocalOverloadState.

File: api/envoy/config/overload/v3/overload.proto
Comment (antoniovicente) [inline api/envoy/config/overload/v3/overload.proto:108]: Should we add some release notes?

**Review by antoniovicente — COMMENTED** [review]
Looks great otherwise.

@ggreenway could you take a look?

File: api/envoy/config/overload/v3/overload.proto
Comment (akonradi) [reply api/envoy/config/overload/v3/overload.proto:108]: Done.

Comment (ggreenway) [reply api/envoy/config/overload/v3/overload.proto:108]: No good deed goes unpunished. There's now a merge conflict on the release notes.

File: include/envoy/event/scaled_timer.h
Comment (ggreenway) [inline include/envoy/event/scaled_timer.h:22]: Please move the definition of all the `operator<<` functions out-of-line (maybe into overload_manager_impl.cc?) to improve compile time.

**Review by ggreenway — CHANGES_REQUESTED** [review]
This looks great, aside from one nit.

File: include/envoy/event/scaled_timer.h
Comment (akonradi) [reply include/envoy/event/scaled_timer.h:22]: These are used implicitly in both //test/common/event:dispatcher_impl_test and //test/server:overload_manager_impl_test. I had added them to make test failures easier to debug (since googletest will print using operator<< or a PrintTo() override), and I don't know that they're actually used anywhere else. Should I just remove them?

Comment (ggreenway) [reply include/envoy/event/scaled_timer.h:22]: If they're unused, then yes, just delete them.

Comment (akonradi) [reply include/envoy/event/scaled_timer.h]: Cool, done.

File: api/envoy/config/overload/v3/overload.proto
Comment (akonradi) [reply api/envoy/config/overload/v3/overload.proto:108]: Funny, it was resolved automatically when I merged main. Fixed now.

Comment (ggreenway) [reply api/envoy/config/overload/v3/overload.proto:108]: Yeah, there's an entry in `.gitattributes` to allow auto-merging adjacent lines, which CLI git will use, but github does not honor: `/docs/root/version_history/current.rst merge=union`

**Review by ggreenway — APPROVED** [review]

**Comment by akonradi** [comment]
@envoyproxy/api-shepherds for the new timer type

**Review by lizan — APPROVED** [review]

**Pull request merged**
